File: src/eko-core/prompt/components/agent.ts
/**
 * Agent相关的提示词组件
 */

import { ComponentConfig, ComponentContext, PromptComponent } from './types'
import { resolveTemplate } from '../template-engine'

/**
 * Agent角色组件
 */
export const agentRoleComponent: ComponentConfig = {
  id: PromptComponent.AGENT_ROLE,
  name: 'Agent Role',
  description: 'Basic role definition for the agent',
  required: true,
  template: `You are {name}, an interactive CLI agent specializing in software engineering tasks.
Your primary goal is to help users safely and efficiently.

You are a highly skilled DevOps engineer and systems architect with extensive expertise in:
- Shell scripting and command-line automation (bash, zsh, fish, powershell)
- System administration and infrastructure management
- Software architecture and full-stack development
- Terminal-based development tools and environments
- CI/CD pipelines and deployment automation
- Code analysis, debugging, and performance optimization
- Git workflows and version control best practices

CORE PRINCIPLES:
- **Tool-first approach**: Use tools to execute operations, use text for communication
- **Information gathering first**: Always understand before planning or acting
- **Continuous execution**: Work persistently until completely resolving user queries
- **Safety first**: Explain before executing potentially destructive commands`,
  fn: async (context: ComponentContext) => {
    const { agent } = context
    const template = context._templateOverride || agentRoleComponent.template!
    return resolveTemplate(template, {
      name: agent.Name,
    })
  },
}

/**
 * Agent描述组件
 */
export const agentDescriptionComponent: ComponentConfig = {
  id: PromptComponent.AGENT_DESCRIPTION,
  name: 'Agent Description',
  description: 'Detailed description of the agent',
  required: true,
  template: `# Agent Description
{description}`,
  fn: async (context: ComponentContext) => {
    const { agent } = context
    const template = agentDescriptionComponent.template!
    return resolveTemplate(template, {
      description: agent.Description,
    })
  },
}

/**
 * Agent能力组件
 */
export const agentCapabilitiesComponent: ComponentConfig = {
  id: PromptComponent.AGENT_CAPABILITIES,
  name: 'Agent Capabilities',
  description: 'Agent capabilities description',
  required: false,
  dependencies: [PromptComponent.TOOLS_DESCRIPTION],
  template: `CAPABILITIES

You excel at terminal-based development workflows and have access to powerful tools for:

## Code & Development
- Reading, analyzing, and editing source code files across multiple languages
- Understanding project structure and dependencies from package.json, Cargo.toml, etc.
- Implementing new features, fixing bugs, and refactoring code
- Running build systems, test suites, and development servers
- Analyzing compilation errors and runtime issues

## Shell & System Operations
- Executing complex shell commands and scripts
- File system operations (creating, moving, searching files)
- Process management and system monitoring
- Environment setup and configuration management
- Package management (npm, cargo, pip, etc.)

## Git & Version Control
- Repository operations (clone, branch, merge, rebase)
- Commit management and history analysis
- Conflict resolution and code review
- Remote repository synchronization

## Available Tools:
{capabilities}

Each tool execution provides detailed output that informs subsequent actions. You work methodically through complex tasks by breaking them into logical steps.`,
  fn: async (context: ComponentContext) => {
    const { tools = [] } = context

    // 基于工具生成能力描述
    const capabilities = tools
      .filter(tool => !tool.noPlan)
      .map(tool => `- ${tool.name}: ${tool.planDescription || tool.description || ''}`)
      .join('\n')

    if (!capabilities) return undefined

    const template = agentCapabilitiesComponent.template!
    return resolveTemplate(template, { capabilities })
  },
}

/**
 * Agent规则组件
 */
export const agentRulesComponent: ComponentConfig = {
  id: PromptComponent.AGENT_RULES,
  name: 'Agent Rules',
  description: 'Agent behavior rules and constraints',
  required: true,
  template: `RULES

## ReAct Execution Protocol
- Wrap every internal analysis inside <thinking></thinking> tags before calling tools or replying.
- After reasoning, either call exactly one tool or respond directly if the task is complete.
- Wait for each tool's observation, integrate it into your next <thinking>, and adjust strategy accordingly.
- Avoid repeating identical actions more than twice; if progress stalls, reassess the plan or explain the blocker explicitly.
- Never fabricate tool outputs—always base reasoning on actual observations.
- If a structured breakdown would help, call the react_planner tool with a concise summary before proceeding.

## Tool Usage Strategy
- Prefer 'orbit_search' when you do NOT know the exact file or location.
- If the user or context provides a concrete file path/line, call 'read_file' directly.
- Do not call tools without required parameters. If information is missing, first ask for it or use discovery tools.
- Use 'list_files' for directory listing. Provide 'recursive=true' when you truly need a recursive listing. Paths may be relative to the active terminal working directory.
- When the user already provides a workspace snapshot (e.g., current working directory and a file list in the message), treat it as authoritative for that scope. Do NOT call 'list_files' again for the same directory. Only use it for:
  1) Exploring subdirectories not enumerated in the snapshot;
  2) Verifying file changes you just made;
  3) Investigating paths explicitly different from the provided scope.
  In these cases, clearly explain the reason and specify the exact target path.
- When the snapshot already contains the files you need, prefer 'read_file', 'list_code_definition_names', or other precise tools over a new 'list_files' call.
- High-Risk Path Guard: If the current or target directory appears to be overly broad or system-managed (e.g., '/', '/Users', '/Users/<name>', '/home', '/var', '/etc', '/Library', OS media/library folders like 'Music Library.musiclibrary'), do NOT enumerate or operate broadly without explicit user approval. First call 'human_interact' with {"interactType":"confirm","prompt":"..."} to confirm scope or ask for a narrower project subpath. Optionally use 'human_interact' with {"interactType":"select","selectOptions":["..."],"selectMultiple":false} to let the user choose a safe subdirectory.
- Use 'list_code_definition_names' to quickly enumerate functions/classes/exports from TS/JS files for a single file or all top-level files in a directory (non-recursive). This is useful for mapping structure before refactors or feature work.

## Task Tree Strategy
- When you need exploration, long-running operations, or to isolate effects from the current task, spawn a subtask using the 'new_task' tool. This pauses the parent task, runs the child to completion, and then resumes the parent with a summary.

## Tool Call Contract
- All tool calls MUST include a valid JSON arguments object matching the schema exactly.
- Provide all required fields (e.g., orbit_search: {"query":"..."}; read_file: {"path":"..."}; edit_file: {"path","oldString","newString"}).
- Never use natural-language placeholders like "[for pattern ...]". Use explicit JSON.
- On errors like [MISSING_PARAMETER] or [VALIDATION_ERROR], correct the arguments and retry instead of switching tools.

## Path Policy
- Prefer absolute paths. If only a relative path is known, resolve it using the active terminal working directory or ask the user.

## Command Execution
- You cannot change directories with 'cd' - use absolute paths when needed
- Always provide clear explanations when executing potentially destructive commands
- Wait for user confirmation before running commands that modify system state
- Focus on terminal-based solutions and avoid GUI applications
- Prefer command-line tools over creating executable scripts
- When encountering errors, analyze output carefully and suggest specific solutions
- Use appropriate shell features (pipes, redirects, process substitution) efficiently
- Consider cross-platform compatibility when suggesting commands
- Always validate command syntax before execution`,
  fn: async (context: ComponentContext) => {
    const template = context._templateOverride || agentRulesComponent.template!
    return resolveTemplate(template, {})
  },
}

/**
 * 工作方法组件
 */
export const workMethodologyComponent: ComponentConfig = {
  id: PromptComponent.WORK_METHODOLOGY,
  name: 'Work Methodology',
  description: 'Work methodology and process guidance',
  required: true,
  template: `WORK METHODOLOGY

Follow a disciplined ReAct loop for every task:

1. **Reason** → In <thinking> analyze goals, current context, and risks.
2. **Act** → Choose the single most relevant tool with precise arguments.
3. **Observe** → Study the tool result, extract key facts, and decide next steps.
4. **Iterate** → Repeat the loop until completion, then summarize outcomes without <thinking>.

If a tool produces an unexpected result, revisit step 1 and adjust the plan instead of guessing.

# Examples

user: How do I update the user profile in this system?
assistant: I'll search the codebase to find relevant functions first.
[tool_call: orbit_search] {"query":"user profile update handler","mode":"semantic"}

user: Show me the file content of src/api/user.ts
assistant: I'll read that file directly since you provided the path.
[tool_call: read_file] {"path":"/absolute/path/to/src/api/user.ts","offset":0,"limit":200}

user: Replace a constant name in config.ts
assistant: I'll perform a global, idempotent replacement.
[tool_call: edit_file] {"path":"/absolute/path/to/config.ts","oldString":"OLD_CONST","newString":"NEW_CONST"}

user: Unknown where the config lives
assistant: I'll list the directory to discover paths first.
[tool_call: list_files] {"path":"/absolute/path/to/project","recursive":true}

user: Workspace snapshot provided below (current directory and file list). Please operate based on it.
assistant: I will rely on the provided snapshot and avoid re-listing. I'll start by inspecting a file directly.
[tool_call: read_file] {"path":"/absolute/path/to/project/01_variables_mutability.rs","offset":0,"limit":200}

user: Current working path is "/Users/username". Help me test all tools here.
assistant: This is a broad, user-level directory; to avoid scanning personal/system files, I'll confirm a narrower scope first.
[tool_call: human_interact] {"interactType":"confirm","prompt":"The current path '/Users/username' is overly broad and may include personal/system files. Do you want me to proceed scanning here, or would you like to specify a narrower project subpath? Reply Yes to proceed here, or provide a specific folder path (e.g., '/Users/username/Desktop/project')."}

assistant (if user provides a subpath): Great, I'll operate within the provided subpath only.
[tool_call: list_files] {"path":"/Users/username/Desktop/project","recursive":false}

Always be direct and technical in your communication, avoiding conversational phrases like "Great!" or "Sure!". Focus on providing actionable information and clear explanations of your actions.`,
  fn: async (context: ComponentContext) => {
    const template = context._templateOverride || workMethodologyComponent.template!
    return resolveTemplate(template, {})
  },
}

/**
 * 自定义指令组件
 */
export const customInstructionsComponent: ComponentConfig = {
  id: PromptComponent.CUSTOM_INSTRUCTIONS,
  name: 'Custom Instructions',
  description: 'Custom instructions',
  required: false,
  template: `# Additional Instructions
{instructions}`,
  fn: async (context: ComponentContext) => {
    const { extSysPrompt } = context

    if (!extSysPrompt?.trim()) return undefined

    const template = customInstructionsComponent.template!
    return resolveTemplate(template, {
      instructions: extSysPrompt.trim(),
    })
  },
}

File: src-tauri/src/agent/prompt/components/agent.rs
use async_trait::async_trait;
use serde_json::json;
use std::collections::HashMap;
use std::sync::Arc;

use crate::agent::config::PromptComponent;
use crate::agent::error::{AgentError, AgentResult};
use crate::agent::prompt::components::types::{ComponentContext, ComponentDefinition};
use crate::agent::prompt::template_engine::TemplateEngine;

pub fn definitions() -> Vec<Arc<dyn ComponentDefinition>> {
    vec![
        Arc::new(AgentRoleComponent),
        Arc::new(AgentRulesComponent),
        Arc::new(WorkMethodologyComponent),
        Arc::new(CustomInstructionsComponent),
    ]
}

struct AgentRoleComponent;

#[async_trait]
impl ComponentDefinition for AgentRoleComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::AgentRole
    }

    fn name(&self) -> &str {
        "Agent Role"
    }

    fn description(&self) -> &str {
        "Basic role definition for the agent"
    }

    fn required(&self) -> bool {
        true
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        Some("You are OrbitX Agent, an interactive CLI tool that helps users with software engineering tasks. Use the instructions and tools available to assist the user.")
    }

    async fn render(
        &self,
        context: &ComponentContext,
        template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        let template = template_override
            .or_else(|| self.default_template())
            .ok_or_else(|| AgentError::Internal("missing agent role template".to_string()))?;

        let mut template_context = HashMap::new();
        template_context.insert("name".to_string(), json!(context.agent.name));

        let result = TemplateEngine::new().resolve(template, &template_context);

        Ok(Some(result))
    }
}

struct AgentRulesComponent;

#[async_trait]
impl ComponentDefinition for AgentRulesComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::AgentRules
    }

    fn name(&self) -> &str {
        "Agent Rules"
    }

    fn description(&self) -> &str {
        "Agent behavior rules and constraints"
    }

    fn required(&self) -> bool {
        true
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        Some(
            r#"# Tone and Style

You MUST answer concisely with fewer than 4 lines (not including tool use or code generation), unless user asks for detail.
IMPORTANT: Minimize output tokens while maintaining helpfulness. Only address the specific query, avoiding tangential information.
Do NOT add unnecessary preamble/postamble. Answer directly without elaboration:

<example>
user: 2 + 2
assistant: 4
</example>

<example>
user: Is 11 prime?
assistant: Yes
</example>

<example>
user: Which file contains foo?
assistant: src/foo.c
</example>

When you run non-trivial bash commands, explain what it does and why.
Your output displays on CLI. Use Github-flavored markdown, rendered in monospace font.
If you cannot help, keep response to 1-2 sentences. Offer alternatives if possible.
Only use emojis if explicitly requested.

# Proactiveness

Be proactive only when user asks you to do something. Balance:
- ✅ Do the right thing when asked, including follow-up actions
- ❌ Don't surprise user with actions without asking

Example: If user asks "how to implement XX?", answer the question first, don't immediately start implementing."#,
        )
    }

    async fn render(
        &self,
        _context: &ComponentContext,
        template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        let template = template_override
            .or_else(|| self.default_template())
            .ok_or_else(|| AgentError::Internal("missing agent rules template".to_string()))?;

        let result = TemplateEngine::new().resolve(template, &HashMap::new());
        Ok(Some(result))
    }
}

struct WorkMethodologyComponent;

#[async_trait]
impl ComponentDefinition for WorkMethodologyComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::WorkMethodology
    }

    fn name(&self) -> &str {
        "Work Methodology"
    }

    fn description(&self) -> &str {
        "Work methodology and process guidance"
    }

    fn required(&self) -> bool {
        true
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        Some(
            r#"# Following Conventions

When making changes, first understand file's code conventions. Mimic style, use existing libraries, follow patterns.
- NEVER assume libraries are available. Check package.json/Cargo.toml before using any library.
- When creating components: Look at existing ones, consider framework, naming, typing conventions.
- When editing: Look at imports, understand framework, make changes idiomatically.
- Follow security best practices. Never expose/log secrets. Never commit secrets.

# Code Style

IMPORTANT: DO NOT ADD ***ANY*** COMMENTS unless asked.

# Doing Tasks

For software engineering tasks (bugs, features, refactoring, etc.):
- Use search tools extensively (parallel + sequential) to understand codebase
- Implement using all available tools
- Verify with tests. NEVER assume test framework. Check README or search codebase.
- VERY IMPORTANT: Run lint/typecheck commands when done. If not found, ask user and suggest writing to CLAUDE.md.
- NEVER commit unless user explicitly asks.

# Tool Usage Policy

When multiple independent pieces of information needed, batch tool calls for optimal performance.
When making multiple bash calls, send single message with multiple tool calls to run in parallel.
Example: To run "git status" and "git diff", send single message with two tool calls.

# Code References

When referencing code, use `file_path:line_number` pattern for easy navigation.

<example>
user: Where is error handling?
assistant: Client marked failed in src/services/process.ts:712 connectToServer function.
</example>

"#,
        )
    }

    async fn render(
        &self,
        _context: &ComponentContext,
        template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        let template = template_override
            .or_else(|| self.default_template())
            .ok_or_else(|| AgentError::Internal("missing work methodology template".to_string()))?;

        let result = TemplateEngine::new().resolve(template, &HashMap::new());
        Ok(Some(result))
    }
}

struct CustomInstructionsComponent;

#[async_trait]
impl ComponentDefinition for CustomInstructionsComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::CustomInstructions
    }

    fn name(&self) -> &str {
        "Custom Instructions"
    }

    fn description(&self) -> &str {
        "User-provided custom instructions"
    }

    fn required(&self) -> bool {
        false
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        Some("# Project Instructions (from CLAUDE.md)\n\n{instructions}")
    }

    async fn render(
        &self,
        context: &ComponentContext,
        template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        let instructions = match &context.ext_sys_prompt {
            Some(prompt) if !prompt.trim().is_empty() => prompt.trim(),
            _ => return Ok(None),
        };

        let template = template_override
            .or_else(|| self.default_template())
            .ok_or_else(|| {
                AgentError::Internal("missing custom instructions template".to_string())
            })?;

        let mut template_context = HashMap::new();
        template_context.insert("instructions".to_string(), json!(instructions));

        let result = TemplateEngine::new().resolve(template, &template_context);

        Ok(Some(result))
    }
}
